sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala

```diff
@@ -221,6 +221,106 @@ object DataSourceAnalysis extends Rule[LogicalPlan] with CastSupport {
   }
 }
 
+/**
+ * Add a repartition before writing Spark SQL Data Sources. It supports three patterns:
+ * 1. Repartition by none when writing normal table/directory.
+ * 2. Repartition by dynamic partition column when writing dynamic partition table/directory.
+ * 3. Repartition by bucket column with bucket number and sort by sort column when writing
+ *    bucket table/directory.
+ *
+ * Note that, this rule must be run after `DataSourceAnalysis`.
+ */
+object RepartitionWritingDataSource extends Rule[LogicalPlan] {
+  override def apply(plan: LogicalPlan): LogicalPlan = {
+    if (conf.repartitionWritingDataSource) {
+      insertRepartition(plan)
+    } else {
+      plan
+    }
+  }
+
+  private def insertRepartition(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
+    case c @ CreateDataSourceTableAsSelectCommand(table, _, query, _) if c.resolved =>
+      val dynamicPartExps = resolveColumnNames(table.partitionColumnNames, query.output)
+      applyRepartition(c, table.bucketSpec, dynamicPartExps)
+
+    case i: InsertIntoHadoopFsRelationCommand if i.resolved =>
```

Comment:
I think we can just use `canApplyRepartition(i.query)` here to simple the code and respect if user repartition by other column.